src/System.IO.FileSystem/src/System/IO/FindEnumerable.Windows.cs

```diff
@@ -0,0 +1,252 @@
+                        {
+                            string subDirectory = PathHelpers.CombineNoChecks(_currentPath, _info->FileName);
+                            IntPtr subDirectoryHandle = CreateDirectoryHandle(subDirectory);
+                            try
+                            {
+                                _pending.Enqueue((subDirectoryHandle, subDirectory));
+                            }
+                            catch
+                            {
+                                Interop.Kernel32.CloseHandle(subDirectoryHandle);
+                                throw;
+                            }
+                        }
+
+                        findData = new RawFindData(_info, _currentPath, _originalFullPath, _originalUserPath);
+                    }
+                } while (!_lastEntryFound && !_predicate(ref findData, _state));
+
+                if (!_lastEntryFound)
+                    _current = _transform(ref findData);
+
+                return !_lastEntryFound;
+            }
+        }
+
+        private unsafe void FindNextFile()
+        {
+            Interop.NtDll.FILE_FULL_DIR_INFORMATION* info = _info;
+            if (info != null && info->NextEntryOffset != 0)
+            {
+                // We're already in a buffer and have another entry
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)((byte*)info + info->NextEntryOffset);
+                return;
+            }
+
+            // We need more data
+            if (GetData())
+                _info = (Interop.NtDll.FILE_FULL_DIR_INFORMATION*)_pinnedBuffer.AddrOfPinnedObject();
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void NoMoreFiles()
+        {
+            _info = null;
+            if (_pending == null || _pending.Count == 0)
+            {
+                _lastEntryFound = true;
+            }
+            else
+            {
+                // Grab the next directory to parse
+                (IntPtr Handle, string Path) next = _pending.Dequeue();
+                Interop.Kernel32.CloseHandle(_directoryHandle);
+                _directoryHandle = next.Handle;
+                _currentPath = next.Path;
+                FindNextFile();
+            }
+        }
+
+        public void Reset()
+        {
+            throw new NotSupportedException();
+        }
+
+        public void Dispose()
+        {
+            Dispose(disposing: true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected void Dispose(bool disposing)
+        {
+            // It is possible to fail to allocate the lock, but the finalizer will still run
+            if (_lock != null)
+            {
+                lock (_lock)
+                {
+                    _lastEntryFound = true;
+
+                    // Don't ever close a valid handle twice as they can be reused- set to zero to ensure this
+                    Interop.Kernel32.CloseHandle(_directoryHandle);
```

Comment:
If Dispose is called twice, won't we end up calling CloseHandle on IntPtr.Zero?

Reply:
Other calls to CloseHandle added in this PR have similar issue as well.

Reply:
Yeah, it is a no-op so I didn't bother checking. The test I added does this.